### ISSUE TITLE
Environment variables are not always set when SSHing into container (from the SSHD script)

### ISSUE THREAD

**Issue by Chuxel** [microsoft/vscode-remote-release#4781]
SSH creates a "login" shell by default (e.g. `bash -li`).  The `-l` flag results in a fresh login shell which pulls environment variables out of /etc/environment, then fires ~/.profile (and by extension ~/.bashrc).

The problem is that the `ENV` directive and `-e` inject environment variables into the container, so there's nothing in classic locations like `/etc/environment` to populate a fresh shell. For example, consider the `julia` image from DockerHub:

```bash
# Interactive shell
$ docker run -it --rm julia /bin/bash -ic 'echo $PATH'
/usr/local/julia/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin

# Login, interactive shell
$ docker run -it --rm julia /bin/bash -lic 'echo $PATH'
/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```

So an interactive shell is safe for largely any image, but login is not. 

The SSHD script in the script library does not solve this problem, so any dev container like codespaces-linux that uses it would see the same thing unless the PATH and other env vars are exxplicitly set in `/etc/profile`, ` /etc/bashrc.bash` or their user specific equivalents.

We need to determine if there is an easy way to fix this in config.

**Comment by Chuxel** [comment]
@chrmarti I've got an updated script in this branch that you can take a look at that is able to restore all env vars in login shells in addition to the `PATH`.  Codespaces secrets are also handled.  https://github.com/microsoft/vscode-dev-containers/blob/clantz/sshd-env-restore/script-library/sshd-debian.sh

The `ENTRYPOINT` script (`/usr/local/share/ssh-init.sh`) saves off enough information to allow a `/etc/profile.d` script to dynamically fix things without overriding things that have been set elsewhere.  I've got some tests that cover a number of scenarios that it seems to cover across sh, bash, and zsh (in another repo, need to merge those into the branch).  This should get things working in the SSH case if they're covered in others.

One question, however. At what point in the process is the PATH merging happening?  For example, if someone ran `docker exec`, (e.g. our friends at Docker) I assume they'd get the unmerged content?

**Comment by chrmarti** [comment]
Relevant discussion: https://github.com/microsoft/vscode-remote-release/issues/5032#issuecomment-849210089

**Comment by chrmarti** [comment]
`/etc/environment` is only used when the sshd config has `UsePAM yes` set. The default is `no`.

SetEnv in the sshd config could be an alternative. (That was added with OpenSSH 7.8 released August 24, 2018.)

`~/.ssh/rc` might be another alternative. (https://man.openbsd.org/sshd_config#PermitUserRC)

The advantage of all 3 is that they do not to depend on a particular shell. /etc/environement depends on PAM, the other two depend on SSH (not sure if there are other types of connections we should consider).

**Comment by Chuxel** [comment]
@chrmarti Good news is `UsePam yes` is set by default on Ubuntu/Debian. This is why the example I pointed to works. https://github.com/chuxel/ssh-env-fix-test

**Comment by Chuxel** [comment]
@chrmarti Yeah anything under `~` is a bit iffy as well since users can get added at any point. It would need to be populated for all of them including root. Not sure if symlinking across users would work. 

The SSH script could look at adding/updating `UsePam yes` to be sure.  Right now it only supports debian/ubuntu, so the Alpine one is N/A at the moment, but clearly other could choose to set it.  In `ratasheep/ubuntu-sshd` its [explicitly disabled](https://github.com/rastasheep/ubuntu-sshd/blob/ed6fffcaf5a49eccdf821af31c1594e3c3061010/18.04/Dockerfile#L12) - likely since this then allows any user to connect rather than a managed list.  I add the specified [user to the ssh group](https://github.com/Chuxel/ssh-env-fix-test/blob/b4fe053fa093c8143753e152228d3aed5bbca40d/.devcontainer/library-scripts/sshd-debian.sh#L74) instead like you would in Ubuntu server / desktop.

**Comment by chrmarti** [comment]
Adding `UsePam yes` works for our base images, but I wouldn't want to do it automatically when creating the dev container. (Still trying to come up with a general solution.)

I see at least two options if someone needs the container env variables in SSH (or some other remote access method):
- Add all container env variables to `/etc/environment` and document to use PAM.
- Add all container env variables to `/etc/profile.d/000-container-env.sh` or similar and document to source that if the user's shell does not use `/etc/profile.d/*`.

The second approach might be more flexible because it doesn't require PAM. It would require the shell to support the `export VAR=VALUE` syntax.

**Comment by Chuxel** [comment]
@chrmarti Yeah I started with `/etc/profile.d`, but I realized that `/etc/bash.bashrc` is sourced before `/etc/profile.d`, so the problem I started seeing was that the you can overwrite variables by mistake or contents in there may depend on these variables. The primary thing I worry about is other path-like variables.

```
if [ "${PS1-}" ]; then
  if [ "${BASH-}" ] && [ "$BASH" != "/bin/sh" ]; then
    # The file bash.bashrc already sets the default PS1.
    # PS1='\h:\w\$ '
    if [ -f /etc/bash.bashrc ]; then
      . /etc/bash.bashrc
    fi
  else
    if [ "`id -u`" -eq 0 ]; then
      PS1='# '
    else
      PS1='$ '
    fi
  fi
fi

if [ -d /etc/profile.d ]; then
  for i in /etc/profile.d/*.sh; do
    if [ -r $i ]; then
      . $i
    fi
  done
  unset i
fi
```

We could document that this is the case, but it started to feel a bit brittle for base env vars.  The alternative would be to inject sourcing an environment file at the very top of `/etc/profile` along with `/etc/zsh/zshenv` since that's common enough to include IMHO.

**Comment by chrmarti** [comment]
Briefly looked into using `SetEnv` in the ssh config. That would work too and seems to be ignored in older versions of sshd (it didn't fail). One disadvantage is that this would override any values set by the user (using `AcceptEnv` or `PermitUserEnvironment`).

Leaning more towards using `/etc/environment` now. It seems you would agree with that?

**Comment by Chuxel** [comment]
@chrmarti Yeah that's where my head is at as well.

**Comment by chrmarti** [comment]
Published Remote-Containers 0.182.0 and Dev Containers CLI 0.0.36 with this fix.
